docs: update docs to include header-based sessions (#556)
supertokens/docs: v2/emailpassword/common-customizations/sessions/token-transfer-method.mdx

---
id: token-transfer-method
title: Token transfer methods
hide_title: true
show_ui_switcher: true
---

<!-- COPY DOCS -->
<!-- ./session/common-customizations/sessions/token-transfer-method.mdx -->

import BackendSDKTabs from "/src/components/tabs/BackendSDKTabs"
import TabItem from '@theme/TabItem';
import NpmOrScriptTabs from "/src/components/tabs/NpmOrScriptTabs"
import AngularUIImplementation from "/src/components/reusableSnippets/angularUIImplementation"
import {Answer}from "/src/components/question"
import {PreBuiltOrCustomUISwitcher, PreBuiltUIContent, CustomUIContent} from "/src/components/preBuiltOrCustomUISwitcher"
import FrontendPreBuiltUITabs from "/src/components/tabs/FrontendPreBuiltUITabs"
import FrontendCustomUITabs from "/src/components/tabs/FrontendCustomUITabs"
import FrontendMobileSubTabs from "/src/components/tabs/FrontendMobileSubTabs"

<!-- TODO: re-think title -->

> # Token transfer methods

Reviewer: How about "Switching between cookies and header based sessions"

Author: That seems a bit too long for the sidebar

Reviewer: It wraps the text on the side bar, so it should be fine

:::caution
We recommend cookie-based sessions in browsers because header-based sessions require saving the access and refresh tokens in storage vulnerable to XSS attacks.
:::

When creating/authorising sessions, the SDK has to choose to send the tokens to the frontend by cookies or custom headers. This choice is ultimately controlled by the backend, but it follows a preference set in the frontend configuration.

> 

Reviewer: I think a better intro is needed:
- SuperTokens supports cookie based and header based tokens. Header based implies that the `authorization bearer` header will be used when making API calls, and that the session tokens can be read from the frontend.
- By default, for browsers, we use cookie based sessions (prevent against token theft via XSS) and for other clients like mobile apps, we use header based.

We should also talk about the motivation for this:
- Mobile devices have buggy cookie handling
- Authorization Bearer tokens make it easier to work with API Gateways and third party services.
- what else??

## Frontend configuration

You can provide a `tokenTransferMethod` property in the configuration of the Session recipe to set the preferred token transfer method, which will be sent to the backend with every request in the `st-auth-mode` header.

By default, the backend follows this preference.

> We recommend using cookie-based sessions in browsers and header-based sessions in mobile apps and environments where you can't or prefer not to use the built-in cookie manager.

Reviewer: ```suggestion
```

<PreBuiltOrCustomUISwitcher>

<PreBuiltUIContent>

<FrontendPreBuiltUITabs>

<TabItem value="reactjs">

```tsx
import SuperTokens from "supertokens-auth-react";
import Session from "supertokens-auth-react/recipe/session";

SuperTokens.init({
    appInfo: {
        apiDomain: "...",
        appName: "...",
        websiteDomain: "..."
    },
    recipeList: [
        Session.init({
            // highlight-next-line

>             tokenTransferMethod: "header"

Reviewer: you should specify all possible values for this - cause some people use JS and they won't get type hint

        })
    ],
})
```

</TabItem>

</FrontendPreBuiltUITabs>

</PreBuiltUIContent>

<CustomUIContent>

<FrontendCustomUITabs>
<TabItem value="web">

<NpmOrScriptTabs>
<TabItem value="npm">

```tsx
import SuperTokens from "supertokens-web-js";
import Session from "supertokens-web-js/recipe/session";

SuperTokens.init({
    appInfo: {
        apiDomain: "...",
        appName: "...",
    },
    recipeList: [
        Session.init({
            // highlight-next-line
            tokenTransferMethod: "header"
        })
    ],
})
```

</TabItem>
<TabItem value="script">

```tsx
import supertokens from "supertokens-web-js-script";
import supertokensSession from "supertokens-web-js-script/recipe/session";
supertokens.init({
    appInfo: {
        apiDomain: "...",
        appName: "...",
    },
    recipeList: [
        supertokensSession.init({
            // highlight-next-line
            tokenTransferMethod: "header",
        })
    ],
})
```

</TabItem>
</NpmOrScriptTabs>

</TabItem>

<TabItem value="mobile">

:::caution
TODO
:::

</TabItem>

</FrontendCustomUITabs>

</CustomUIContent>

</PreBuiltOrCustomUISwitcher>

## Backend configuration

:::important

> By returning a fixed value from `getTokenTransferMethod` you can strictly control the allowed session types, but changing this value can result in logging out users.

Reviewer: ```suggestion
By returning a fixed value from `getTokenTransferMethod` you can strictly control the allowed session types regardless of the frontend config.
```

:::

> You can provide a `getTokenTransferMethod` callback in the configuration of the Session recipe. This callback controls:
> - The type of sessions `createNewSession` and the login APIs create.
> - The types of sessions allowed when verifying sessions.

Reviewer: ```suggestion
You can provide a `getTokenTransferMethod` callback in the configuration of the Session recipe.
```

> By default, we allow both cookie and header sessions during authorisation and follow the preference of the frontend while creating new sessions. The token transfer method of the frontend is attached to requests in the `st-auth-mode` header.

Reviewer: this can go as a `:::note` at the end of the page

> By default, we allow both cookie and header sessions during authorisation and follow the preference of the frontend while creating new sessions. The token transfer method of the frontend is attached to requests in the `st-auth-mode` header.

Reviewer: ```suggestion
By default, we allow both cookie and authorization bearer tokens during session verification. When creating a new session, we follow the preference of the frontend indicated by the `st-auth-mode` request header (set by the frontend SDK).
```

        Session.init({
            // highlight-next-line
            tokenTransferMethod: "header" // or "cookie"
        })
    ],
})
```

</TabItem>

</FrontendPreBuiltUITabs>

</PreBuiltUIContent>

<CustomUIContent>

<FrontendCustomUITabs>
<TabItem value="web">

<NpmOrScriptTabs>
<TabItem value="npm">

```tsx
import SuperTokens from "supertokens-web-js";
import Session from "supertokens-web-js/recipe/session";

SuperTokens.init({
    appInfo: {
        apiDomain: "...",
        appName: "...",
    },
    recipeList: [
        Session.init({
            // highlight-next-line
            tokenTransferMethod: "header" // or "cookie"
        })
    ],
})
```

</TabItem>
<TabItem value="script">

```tsx
import supertokens from "supertokens-web-js-script";
import supertokensSession from "supertokens-web-js-script/recipe/session";
supertokens.init({
    appInfo: {
        apiDomain: "...",
        appName: "...",
    },
    recipeList: [
        supertokensSession.init({
            // highlight-next-line
            tokenTransferMethod: "header" // or "cookie",
        })
    ],
})
```

</TabItem>
</NpmOrScriptTabs>

</TabItem>

<TabItem value="mobile">

:::caution
TODO

> :::

Reviewer: TODO for mobile SDKs
